### DISCUSSION TITLE
How to add periodic events and schedule them with this app?

### DISCUSSION THREAD

**Discussion by newbeelearn** [pocketbase/pocketbase#245] [Q&A] [answered]
There are certain tasks that need to happen in periodic intervals without any trigger from outside for example checking daily stock price. This kind of task desn't requires any input from user. Is there any way to do this in the current app? if not is there any plan to support it in future or does it looks outside of the scope? what is the guidance on this?

**Comment by ganigeorgiev** [comment, marked as answer]
>  Is there any way to do this in the current app?

Out of the box - No. But PocketBase can be used as Go framework/toolkit and  you are free to implement whatever custom business logic you want, so for example you could easily add a timer with callback or 3rd-party cron-like library to execute your task(s).

Similar to the previous aggregations and reporting feature requests, there isn't really a clear definition of what a user task could be and therefore the implementation will differ from use case to use case, or in other words - it is too "custom" to be integrated easily in a generic/common way in the prebuilt executable.

One possible solution, so that users would not be required to write so much custom app code, could be to implement a "Workflow/Triggers builder" with various configurable "plugins/blocks" (like send email on some event, fetch and populate data, etc.), but, considering that I have another open source project to maintain + my regular job, I don't think that I'll be able to work on this (at least for v1.0.0).

**Reply by newbeelearn** [reply]
I can help with the code in case you plan to add this feature. It would still put some burden on your time as you would have to review the changes and clarify the design decisions also i am bit new to web development so i am not sure of my code quality.

**Reply by ganigeorgiev** [reply]
Contributions are more than welcome, but  implementing something like a "Workflow/Triggers builder" requires a lot of planning and really careful design considerations (both in terms of code and UX). So if you want to work on this, please outline a proposal and share your idea before proceeding to prevent wasting your time.

**Reply by newbeelearn** [reply]
Ok, i am thinking of adding cron library to the code base[1] . This is popular library and doesn't have any dependency. For now cron would be limited to adding "trigger" to the already existing add(fn) and trigger(data) so it would only be useful to the people who are using this as framework.

For workflow i want to add new field type called trigger in collection with the fields like once/periodic, interval and a url to the request to so actual action would not be part of the pocketbase instead it would be limited to calling external url with action. This would help in adding zapier/lambda/worker etc. without adding much burder of providing actions/plugins within the pocketbase. This would be in second phase.

what do you think?

[1] https://github.com/robfig/cron

**Reply by ganigeorgiev** [reply]
1. Why do you want the cron library to be part of the PocketBase source?
  I don't think creating a wrapper for the cron library would help with the dev experience and actually could make it worst because the developers will have to be aware of both the PocketBase specific wrapper implementation and the original library behavior.

2. About the "trigger" field:
- Collection fields are represented as db columns in the underlying collection records table. How this filed will be stored? Are you sure that you don't want to add the "trigger" as collection setting not as field that could have different value for each record?
- If the latter - how this field will be set? Only from the Admin? Should we add special handling when the users have create/update permissions for the records?
- Currently all collection fields could take part in the API rules and filters. How the trigger field would like as part of such rule or user defined filter?
- Which part of the code will be responsible for calling the trigger?
- How the developers who uses PocketBase as framework will be able to control the trigger behavior (eg. intercept, overwrite)?
  
  From the description, it seems that you are looking for Webhooks-like behavior. I don't think implementing it as Collection field would be the best approach. And even if we implement it as collection setting(s) it will very limited - what if some users want to send additional data to the webhook endpoint, to customize the HTTP verb (GET, POST, DELETE, etc.), or to send to multiple urls at once, etc. If we are not going to implement workflow actions/plugins, in my opinion it would be a lot easier and flexible for the developer to just use the existing code event hooks and to call on their own the endpoints they want to notify with the standard Go http client.

**Reply by newbeelearn** [reply]
>  If we are not going to implement workflow actions/plugins, in my opinion it would be a lot easier and flexible for the developer to just use the   existing code event hooks and to call on their own the endpoints they want to notify with the standard Go http client.

Ok, thanks for the quick response. As you suggested I will use http client for  now lets see how it goes.

**Reply by mbecker** [reply]
Hi,
I'm using Pocketbase as a framework and extending some functionalities for my special use case like sending Kafka message for DB inserts / updates or in this case a cronjob to refresh the access token from the oauth provider. I'm using the library `github.com/robfig/cron/` to run a job every minute as follows:

```go
...
c := cron.New()
c.AddFunc("*/1 * * * *", func() {
			nowTime := time.Now().UTC().Add(30 * time.Minute).Format(types.DefaultDateLayout)
			log.Printf("Starting cronjob: OAuthToken refresh with time=%s", nowTime)
			oAuthTokenCollection, err := app.Dao().FindCollectionByNameOrId(migrations.OAuthTokenCollectionName)
			if err != nil {
				log.Println(err)
				return
			}
			expr := dbx.NewExp("expiry<{:now}", dbx.Params{"now": nowTime})
			oAuthTokenRecords, err := app.Dao().FindRecordsByExpr(oAuthTokenCollection, expr)
			if err != nil {
				log.Println(err)
				return
			}
...
```

Works for me perfectly. Hope this helps.

**Reply by ganigeorgiev** [reply]
@newbeelearn Just to let you know that I've created an issue for integrating a builtin webhook and mail triggers in the Admin UI. You could review the proposal and share your feedback in https://github.com/pocketbase/pocketbase/issues/317.
